### PR TITLE
[#139] Add auxiliary input type 2 support

### LINKED ISSUE #139: Auxiliary input type 2 support
Hi, right now there is support for the so called 'function' part of AUX-N. These would be used by the ECU controlling the implement behind the tractor. However, these functions are driven by the buttons/switches inside the tractor which are called the 'input' part of AUX-N. To support the input part from within the stack, I think we have to implement the following:

- [ ] VT Status (Learn Mode)
- [ ] Auxiliary Input Status Type 2 Enable command (VT -> ECU)
- [ ] Auxiliary Input Status Type 2 Enable response (ECU -> VT)
- [ ] Auxiliary Input Status message:
  Iff Learn Mode is active: ECU -> VT, otherwise ECU -> Broadcast
- [ ] Auxiliary Input Maintenance message (ECU -> Broadcast)

When this is implemented we can possibly do unit/integration tests with both of them talking to each other?

### PR DESCRIPTION
**Pull request by GwnDaan**
**What's new:**
- Implementation of aux-n input handling from within the stack (See #139 for details)
- Added example for aux-n input application (dir: `examples/vt_aux_n/inputs`)
- Moved aux-n function example application to own folder (dir: `examples/vt_aux_n/functions`)

Note: we probably want to wait for #65 to be finished before merging this PR. Right now a `auxiliary input maintenance message` is send every 100ms, regardless of whether there is an auxiliary input present in the object pool. A few other things also should rely on an object being in the object pool. I could've made some configurable options for this, but I'd rather wait so we can do it right the first time. I'd love to hear any opinions on this.

Closes #139

### CODE REVIEW

File: examples/vt_aux_n/inputs/main.cpp
Comment (GwnDaan) [inline examples/vt_aux_n/inputs/main.cpp:179]: Should be removed once we can derive which input ObjectID's are there from the uploading object pool process

File: isobus/src/isobus_virtual_terminal_client.cpp
Comment (GwnDaan) [inline isobus/src/isobus_virtual_terminal_client.cpp:313]: We should do the adding of new inputs to `ourAuxiliaryInputs` when parsing/uploading the object pool

File: isobus/src/isobus_virtual_terminal_client.cpp
Comment (GwnDaan) [inline isobus/src/isobus_virtual_terminal_client.cpp:1982]: We can actually change this to setting the flag only when `ourAuxiliaryInputs` is not empty I think

**Review by GwnDaan — COMMENTED** [review]
The related TODO's to #65 which I believe we should do before merging this PR

**Comment by GwnDaan** [comment]
Gotta love delaying my PR's haha, so many things have changed since the initial commits... And the rebase couldn't resist leaving some artifacts behind as well. Seems like I'll have to spend some time this weekend fixing this 😅

File: isobus/src/isobus_virtual_terminal_client.cpp
Comment (GwnDaan) [reply isobus/src/isobus_virtual_terminal_client.cpp]: Instead of parsing them from the object_pool (probably resource consuming and complicated task), you have to register the input object ids manually with setters.

**Comment by ad3154** [comment]
Let me know when you think this is ready for review and I'll go through it 🙂

**Comment by GwnDaan** [comment]
> Let me know when you think this is ready for review and I'll go through it 🙂

I think it's ready for review. I still have to write some tests, but the rest is finished

File: examples/virtual_terminal/aux_inputs/main.cpp
Comment (ad3154) [inline examples/virtual_terminal/aux_inputs/main.cpp:23]: I think the comments on these two values might be reversed

File: examples/virtual_terminal/aux_inputs/main.cpp
Comment (ad3154) [inline examples/virtual_terminal/aux_inputs/main.cpp:46]: If this commented out code is just for debugging the example we may want to remove it, same goes for some other spots in this file.
Re: Autosar Rule A2-8-2 Sections of code shall not be “commented out”.

File: examples/virtual_terminal/aux_inputs/main.cpp
Comment (ad3154) [inline examples/virtual_terminal/aux_inputs/main.cpp:109]: Mostly a nit, if you are `using namespace std` I'd suggest either omitting `std::` everywhere or removing `using namespace std`

File: isobus/src/isobus_virtual_terminal_client.cpp
Comment (ad3154) [inline isobus/src/isobus_virtual_terminal_client.cpp:2468]: I think this function can be `const`

File: isobus/src/isobus_virtual_terminal_client.cpp
Comment (ad3154) [inline isobus/src/isobus_virtual_terminal_client.cpp:2486]: I think this function can be `const`?

File: isobus/src/isobus_virtual_terminal_client.cpp
Comment (ad3154) [inline isobus/src/isobus_virtual_terminal_client.cpp:2518]: Can you newline some of these conditions? Having each on their own line is nice

File: isobus/src/isobus_virtual_terminal_client.cpp
Comment (ad3154) [inline isobus/src/isobus_virtual_terminal_client.cpp:3570]: If this is true, should we `break;` below?

File: isobus/src/isobus_virtual_terminal_client.cpp
Comment (ad3154) [inline isobus/src/isobus_virtual_terminal_client.cpp:3203]: Should we use `std::find_if` ? I don't have a lot of experience with it personally but situations like this and line 3120 make me think "hmmmm"

**Comment by GwnDaan** [comment]
I'll try to go over your comments later today, having a lot less spare time lately unfortunately... Other than issue #192, there were not really any issues while testing this thoroughly. But I don't really have the time to write unit tests any time soon I'm afraid.

File: isobus/src/isobus_virtual_terminal_client.cpp
Comment (GwnDaan) [reply isobus/src/isobus_virtual_terminal_client.cpp:3203]: Isn't `std::find_if` a C++20 feature? https://en.cppreference.com/w/cpp/algorithm/find

Comment (ad3154) [reply isobus/src/isobus_virtual_terminal_client.cpp:3203]: Although it's definitely in 20, it seems like there might be a version prior as well, as theres a "until C++20" on one of them?
image

**Comment by ad3154** [comment]
> I'll try to go over your comments later today, having a lot less spare time lately unfortunately... Other than issue #192, there were not really any issues while testing this thoroughly. But I don't really have the time to write unit tests any time soon I'm afraid.

No worries, I totally understand! I have been quite busy recently as well. We can always come back and make incremental improvements and add tests later too.

File: isobus/src/isobus_virtual_terminal_client.cpp
Comment (GwnDaan) [reply isobus/src/isobus_virtual_terminal_client.cpp:3570]: Switched to `std::find_if`

File: isobus/src/isobus_virtual_terminal_client.cpp
Comment (GwnDaan) [reply isobus/src/isobus_virtual_terminal_client.cpp:3203]: Ah, I see. I like `find_if` more now I used it a couple of times 😄

File: examples/virtual_terminal/aux_inputs/main.cpp
Comment (GwnDaan) [reply examples/virtual_terminal/aux_inputs/main.cpp:109]: Removed `using namespace std` from all examples as we never omitted `std::` by the looks of it

File: examples/virtual_terminal/aux_inputs/main.cpp
Comment (GwnDaan) [reply examples/virtual_terminal/aux_inputs/main.cpp:23]: Good catch! Fixed in latest commit

File: examples/virtual_terminal/aux_inputs/main.cpp
Comment (GwnDaan) [reply examples/virtual_terminal/aux_inputs/main.cpp:105]: Actually nevermind, for some reason my windows machine decided building was fine while it clearly isn't

Comment (GwnDaan) [reply examples/virtual_terminal/aux_inputs/main.cpp:105]: `std::` was only not present for some of the `cout` and `endl` in the transport_layer example, so I added it instead of adding back the `using namespace std`. Though I'll change it back happily if the opposite is desired

**Comment by GwnDaan** [comment]
> No worries, I totally understand! I have been quite busy recently as well. We can always come back and make incremental improvements and add tests later too.

Yea, I think this PR can be merged if you think it's ready, I'll try to add tests someday but that won't be anytime soon

**Comment by ad3154** [comment]
@GwnDaan When you get time, if you could fix the conflicts in those 2 files (CMakeLists.txt and test/vt_client_tests.cpp) then I'll get this merged in before it sits around any longer (haha, sorry about that)

**Comment by GwnDaan** [comment]
> @GwnDaan When you get time, if you could fix the conflicts in those 2 files (CMakeLists.txt and test/vt_client_tests.cpp) then I'll get this merged in before it sits around any longer (haha, sorry about that)

Should be fixed now 😄

**Review by ad3154 — APPROVED** [review]

**Pull request merged**
